sys/kern: Enable support for rv64

furquan-rivos opened the pull request:
This change copies riscv32.rs to riscv64.rs under sys/kern/src/arch and updates for rv64. Things that are dropped from rv64 for now are mtimer and plic support. PMP configuration doesn't take into account that different DRAM regions can have different access permissions.

furquan-rivos committed:
sys/kern: Enable support for rv64

This change copies riscv32.rs to riscv64.rs under sys/kern/src/arch
and updates for rv64. Things that are dropped from rv64 for now are
mtimer and plic support. PMP configuration doesn't take into account
that different DRAM regions can have different access permissions.

FawazTirmizi commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -0,0 +1,603 @@
+        Trap::Exception(Exception::InstructionFault) => unsafe {
+            handle_fault(task, FaultInfo::IllegalText);
+        },
+        _ => {}
```
This will also ignore any exceptions that aren't accounted for above. Maybe instead check if it's an exception and return only if it isn't?

furquan-rivos replied:
My plan is to bring back `klog` for kernel and enable some useful prints around such cases that are not handled. Currently, this panics on rv32 which doesn't really generate any log (at least when humility is not connected). So adding a panic wouldn't really help much anyways. I am tracking the activity to bring back klog and add trap handling messages in https://rivosinc.atlassian.net/browse/SW-479. Does that plan look okay?

FawazTirmizi replied:
Having a panic would at least keep the kernel in an infinite loop, so the state of the system will be preserved, which would make it easier to trace the cause of the exception. Unless we also plan to allow the system to keep running on arbitrary exceptions? Either way, if you think it's unnecessary that's enough for me.

furquan-rivos replied:
Added a panic for now in a follow-up commit.

FawazTirmizi commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -0,0 +1,603 @@
+    Timestamp::from(unsafe { TICKS })
+}
+
+#[allow(unused_variables)]
+pub fn disable_irq(n: u32) {}
+
+#[allow(unused_variables)]
+pub fn enable_irq(n: u32) {}
```
May be off-topic, but is the plan to leave notifications as 32-bit for this port?

furquan-rivos replied:
Yes, I don't think we would need more than 32 notifications per task. So, for now leaving this at 32-bit seems fine.

FawazTirmizi replied:
Sounds good.

FawazTirmizi reviewed (commented):
Besides the comment on exception handling, LGTM.

furquan-rivos committed:
sys/kern/rv64: Panic in case of unhandled expection

This change adds a panic if we encounter a trap that is not handled.

beezow commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -0,0 +1,603 @@
+        self.a5 as usize
+    }
+    fn arg6(&self) -> usize {
+        self.a6 as usize
```
nit: Why are the `arg` registers usize, while the `ret` registers u64?  I know they will end up being the same, but just wondering if it was chosen for a reason.

beezow replied:
It seems like a lot of the changes are switching from u32 to u64.  Could we avoid diverging this file by using usize?  Aside from that there are only a couple of changes and they use the bytes per word. I think that could just be a static const selected at compile time.

furquan-rivos replied:
> nit: Why are the arg registers usize, while the ret registers u64? I know they will end up being the same, but just wondering if it was chosen for a reason.

Params for both `argX()` and `retX()` methods are usize because it implements the [ArchState trait](https://github.com/rivosinc/hubris/blob/rivos/main-rv64/sys/kern/src/task.rs#L353). The `ArchState` trait uses usize because it is implemented by different architectures - 32 and 64bit.

For RV64, these registers are all fixed-width. Hence, `SavedState` uses `u64` and does the translation using `as usize` and `as u64` in the respective methods.

> It seems like a lot of the changes are switching from u32 to u64. Could we avoid diverging this file by using usize?

Do you mean diverging from riscv32.rs? I think its best to keep RV32 and RV64 support separate. There are going to be many more changes coming up that would probably take RV32 and RV64 further apart. We can always bring them back together into a common file if we realize later that they aren't really that different after all.

beezow replied:
Sounds good.

beezow commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -0,0 +1,603 @@
+        register::mie::set_mtimer();
+
+        // Global machine interrupt enable
+        register::mstatus::set_mie();
```
It looks like you inherited a bug from an earlier version of hubris.  This should be a `set_mpie()`.  Once `MEPC` is set, any interrupt before the mret will crash the boot sequence.

furquan-rivos replied:
Setting `MPIE` would set previous-mode's interrupt enable bit on mret. However, at this point, the aim is to enable interrupts for machine mode. I don't think I understood what bug you are referring to. Is the bug because `CURRENT_TASK_PTR` is not set? If so, then `CURRENT_TASK_PTR` can be set before enabling machine interrupts. But, the intent here was to enable machine interrupts.

furquan-rivos replied:
I think I see what you are saying. We don't really want machine-mode interrupts to be enabled while we are still in M-mode. We are basically running without interrupts enabled in M-mode all the time. Is there any benefit of setting MPIE at all for us? We don't really support nested interrupts and interrupts to M-mode are default enabled when going to lower modes (S/U-mode).

beezow replied:
So we don't actually want to enable interrupts yet.  On line 525 we set `mepc` so it is ready when we `mret` to the first task.  However, any interrupt will change mepc.  If we instead just set `mpie`, after we `mret`, all interrupts will be enabled.  
Aside from this, I think interrupts are always disabled when in m-mode for hubris.  The only way to get into m-mode is through the trap handler, which disables on entry, and re enables on exit.

furquan-rivos replied:
Done. Pushed a change to fix this: https://github.com/rivosinc/hubris/pull/17/commits/0550dbb00f9d70bbbc45cb63b189461f4d267601

furquan-rivos replied:
Looks like I missed your recent response. 

>  If we instead just set mpie, after we mret, all interrupts will be enabled.

We don't really need to set MPIE as well. Interrupts to M-mode are default enabled in all lower modes.

beezow replied:
I saw you latest push.  I forget we are not delegating interrupts so MPIE doesn't really matter here.  I didn't realize m-mode interrupts were always enables when in a lower privilege mode either.  Will make a patch up for the main rv32 port to correct this.

beezow commented on sys/kern/src/arch/riscv64.rs:
```diff
@@ -0,0 +1,603 @@
+        if (region.base == 0x0) && (region.size == 0x20) {
+            continue;
+        }
+        let pmpcfg = {
+            let pmp_perm: Permission = match region.attributes.bits() & 0b111 {
+                0b000 => Permission::NONE,
+                0b001 => Permission::R,
+                0b010 => panic!(),
+                0b011 => Permission::RW,
+                0b100 => Permission::X,
+                0b101 => Permission::RX,
+                0b110 => panic!(),
+                0b111 => Permission::RWX,
+                _ => unreachable!(),
+            };
```
Should this be a operation within the riscv crate?  I think converting RWX permissions to a bitmask is pretty standard?
@FawazTirmizi

furquan-rivos replied:
Probably. But out of scope for this change, I think. We can update this function once the riscv crate is updated.

furquan-rivos replied:
Actually, thinking about this some more. The region attributes don't really mean anything for the riscv crate. So, I don't know if it really makes sense to add region permissions -> pmp cfgs in the riscv crate directly.

furquan-rivos committed:
sys/kern/rv64: Drop setting of MIE

We do not support interrupts while running in M-mode. As per RV Priv Spec,
M-mode interrupts are enabled by default when running in any lower privileged
mode. Thus, this change drops the enabling of MIE completely.

TEST=Verified that all tests are still passing as before for rv64
